chainer/_link.py

```diff
@@ -0,0 +1,373 @@
+            skip_between_cupy_devices=True)
+        self.__to_device(visitor)
+        return self
+
+    def to_intel64(self):
+        # type: () -> 'DeviceResidentMixin'
+        """Copies parameter variables and persistent values to CPU."""
+        intel64.check_ideep_available()
+        visitor = _ToDeviceVisitor(
+            chainer.get_device(intel64),
+            entry_method_info=('to_intel64', {}))
+        self.__to_device(visitor)
+        return self
+
+    def to_chx(self):
+        """Converts parameter variables and persistent values to ChainerX \
+without any copy.
+
+        This method does not handle non-registered attributes. If some of such
+        attributes must be copied to ChainerX, the link implementation must
+        override this method to do so.
+
+        Returns: self
+        """
+        if not chainerx.is_available():
+            raise RuntimeError('ChainerX is not available.')
+
+        if self.xp is chainerx:
+            return self
+
+        self._device = None
+        self.visit_device_residents(_ToChxVisitor())
+        return self
+
+    def from_chx(self):
+        """Converts parameter variables and persistent values from ChainerX \
+to NumPy/CuPy devices without any copy."""
+        if isinstance(self._device, backend.ChainerxDevice):
+            self._device = self._device.fallback_device
+
+        self._device = None
+        self.visit_device_residents(_FromChxVisitor())
+        return self
+
+    def __to_device(self, to_device_visitor):
+        self._device = None
+        self.visit_device_residents(to_device_visitor)
+
+    def to_device(
+            self,
+            device  # type: types.DeviceSpec
+    ):
+        # type: (...) -> 'DeviceResidentMixin'
+        """Copies parameter variables and persistent values to the specified \
+device.
+
+        This method does not handle non-registered attributes. If some of such
+        attributes must be copied to the device, the link implementation must
+        override this method to do so.
+
+        Args:
+            device: Target device specifier. See
+                :func:`~chainer.get_device` for available values.
+
+        Returns: self
+
+        """
+        device = chainer.get_device(device)
+        self.__to_device(_ToDeviceVisitor(device))
+        return self
+
+
+def _is_to_device_method_overridden(device_resident, method_name):
+    # Returns whether the specified to_device family method is overridden.
+    to_method = getattr(device_resident, method_name, None).__func__
+    to_method_orig = getattr(DeviceResidentMixin, method_name)
+    if sys.version_info < (3,):
+        to_method_orig = to_method_orig.__func__
+    if to_method is not to_method_orig:
+        return True  # overridden
+    return False
+
+
+class DeviceResidentsVisitor(object):
+    # Base class of visitors.
+
+    def visit_visitable(self, visitable, visitor):
+        assert isinstance(visitable, DeviceResidentMixin)
```

Comment:
If the type hierarchy means something, I don't think it is good to call the superclass "mixin". How about just naming it `DeviceResident`?

Reply:
Fixed.